### PR TITLE
issue-752 Atlas security - clone and modify roles

### PR DESCRIPTION
**Pull request by YaroslavTir**
Fixes https://github.com/OHDSI/WebAPI/issues/752
 add json export/import functionality for roles

### CODE REVIEW

File: js/pages/configuration/roles/role-import.html [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 <div data-bind="css: classes(), if: isAuthenticated() && hasAccess()">
   <div data-bind="css: classes('source')">
-    <div data-bind="css: classes('json-header-wrapper')">
+    <div data-bind="css: classes('json-header-wrapper')">isJSONValid
```
Comment (pavgra) [inline js/pages/configuration/roles/role-import.html]: Why the `isJSONValid` is here?

File: js/pages/configuration/roles/role-details.js [hunk 2]
```diff
@@ -109,36 +127,75 @@ define([
...
-            const role = await roleService.load(this.roleId());
-            this.roleName(role.role);
-        }
 
-        async getUsers() {
-            if (!this.users() || this.users().length == 0) {
-                const users = await userService.getUsers();
-                this.users(users);
+        reload () {
+            if (this.modifiedJSON().length > 0 && this.isJSONValid) {
```
Comment (pavgra) [inline js/pages/configuration/roles/role-details.js]: Shouldn't the `this.isJSONValid` be a function call? (`this.isJSONValid()`)

File: js/pages/configuration/roles/role-details.js [hunk 3]
```diff
@@ -179,6 +236,73 @@ define([
...
+                permissions: this.getPermissionsList().map(p => ({ id: p.permission() })),
+                users: this.getUsersList().map(u => ({ id: u.login })),
+            }];
+
+            fileService.saveAsJson(role);
+        }
+
+
+        copyRoleExpressionJSONToClipboard () {
+            this.copyToClipboard('#btnCopyExpressionJSONClipboard', '#copyRoleExpressionJSONMessage');
```
Comment (pavgra) [inline js/pages/configuration/roles/role-details.js]: I would move this into constants

File: js/pages/configuration/roles/role-details.html [hunk 4]
```diff
@@ -25,8 +25,15 @@
             </div>
         </div>
         <ul class="nav nav-tabs">
-            <li role="presentation" data-bind="click: function() {$component.currentTab('users');}, attr: {class: $component.currentTab() == 'users' ? 'active' : ''}"><a>Users</a></li>
-            <li role="presentation" data-bind="click: function() {$component.currentTab('permissions');}, attr: {class: $component.currentTab() == 'permissions' ? 'active' : ''}"><a>Permissions</a></li>
+            <li role="presentation" data-bind="click: function() {$component.currentTab('users');}, attr: {class: $component.currentTab() == 'users' ? 'active' : ''}">
```
[the remark this answers is not in the document]
Comment (YaroslavTir) [reply js/pages/configuration/roles/role-details.html]: hmm, do you mean use <tabs> tag and put each tav-div into a separate page?

File: js/pages/configuration/roles/role-details.js [hunk 5]
```diff
@@ -109,36 +130,75 @@ define([
...
-            this.roleName(role.role);
-        }
 
-        async getUsers() {
-            if (!this.users() || this.users().length == 0) {
-                const users = await userService.getUsers();
-                this.users(users);
+        reload () {
+            if (this.modifiedJSON().length > 0 && this.isJSONValid()) {
+                let object = JSON.parse(this.modifiedJSON());
```
[the remark this answers is not in the document]
Comment (YaroslavTir) [reply js/pages/configuration/roles/role-details.js]: fixed

File: js/pages/configuration/roles/role-details.js [hunk 6]
```diff
@@ -179,6 +239,73 @@ define([
         }
 
 
+        getPermissionsList() {
+            return this.permissionItems()
+              .filter(permission => permission.isRolePermission());
+        }
+
+        getUsersList() {
```
[the remark this answers is not in the document]
Comment (YaroslavTir) [reply js/pages/configuration/roles/role-details.js]: there is no really strong reason behind this, I just put methods in a more logical way. is it critical? as I see only 3 small methods are affected `getPermissionsList`,  `getUsersList`, `export`

**Comment by vlbe** [comment]
@YaroslavTir 
Maybe we should make such functionality as in other places(IR/Cohort Pathways/PLP/PLE/Characterization/etc):
- Add new tab 'Utilities'.
- Move import and export functionality to this tab.

File: js/pages/configuration/roles/components/permissions.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+            super(params);
+
+            this.isNewRole = params.isNewRole;
+            this.roleId = params.roleId;
+            this.permissionItems = params.permissionItems;
+
+            this.canEditRolePermissions = ko.pureComputed(() => { return authApi.isAuthenticated() && (this.isNewRole() || authApi.isPermittedEditRolePermissions(this.roleId())); });
+        }
+
+        renderCheckbox(field, editable) {
```
Comment (vlbe) [inline js/pages/configuration/roles/components/permissions.js]: I think we already have some function in commonUtils to renderCheckbox

File: js/pages/configuration/roles/components/users.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+        selectAllUsers() {
+            this.userItems().forEach(user => user.isRoleUser(true));
+        }
+
+        deselectAllUsers() {
+            this.userItems().forEach(user => user.isRoleUser(false));
+        }
+
+        renderCheckbox(field, editable) {
```
Comment (vlbe) [inline js/pages/configuration/roles/components/users.js]: same as above

**Pull request merged**